Added memoization in expiry time in some places

manojmalik20 opened the pull request:
Added memoization in status dropdown and profile popover
Added logic to show tooltip in custom status emoji in channel header
Refactored some code in custom status emoji and custom status selector
Added timezone dependent current time in custom status selector

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has redux changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```release-note
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
```

manojmalik20 committed:
Added memoization in expiry time in some places
Added memoization in status dropdown and profile popover
Added logic to show tooltip in custom status emoji in channel header
Refactored some code in custom status emoji and custom status selector
Added timezone dependent current time in custom status selector

manojmalik20 committed:
Added unit tests and several fixes
Sorted the order of imports in several files
Made unit tests for date_time_input and expiry_menu components
Updated some snapshots

manojmalik20 committed:
Fixed custom status emoji unit tests

chetanyakan committed:
Fix types

chetanyakan committed:
Merge branch 'MI-1235' of github.com:brightscout-alpha/mattermost-webapp into MI-1235

* 'MI-1235' of github.com:brightscout-alpha/mattermost-webapp:
  Fixed custom status emoji unit tests

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -177,12 +177,12 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
     };
 
-    let disableSetStatus = (currentCustomStatus.text === text && currentCustomStatus.emoji === emoji) ||
+    let disableSetStatus = (currentCustomStatus?.text === text && currentCustomStatus.emoji === emoji) ||
```
rename to `isCustomStatusEmojiAndTextSame`

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -177,12 +177,12 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
     disableSetStatus = Boolean(disableSetStatus &&
-        ((expiry !== CustomStatusDuration.DATE_AND_TIME && currentCustomStatus.duration === expiry) ||
-            (expiry === CustomStatusDuration.DATE_AND_TIME && currentCustomStatus.expires_at && customExpiryTime.getTime() === new Date(currentCustomStatus.expires_at).getTime())));
+        ((expiry !== CustomStatusDuration.DATE_AND_TIME && currentCustomStatus?.duration === expiry) ||
+            (expiry === CustomStatusDuration.DATE_AND_TIME && currentCustomStatus?.expires_at && customExpiryTime.getTime() === new Date(currentCustomStatus.expires_at).getTime())));
```
nit - if possible, refactor this to multiple boolean variables to facilitate ease of readability

chetanyakan commented on components/profile_popover/profile_popover.jsx:
```diff
@@ -299,7 +301,7 @@ class ProfilePopover extends React.PureComponent {
             expiryContent = customStatusSet && customStatus.expires_at && customStatus.duration !== CustomStatusDuration.DONT_CLEAR && (
                 <span>
                     {' (Until '}
-                    {displayExpiryTime(customStatus.expires_at, this.props.timezone)}
+                    {this.displayExpiryTime(customStatus.expires_at, this.props.timezone)}
                     {')'}
                 </span>
             );
```
If this is repeated a couple times, maybe you can refactor this to a common component

manojmalik20 committed:
Refactored some code
Converted displayExpiryTime util to a separate functional component
Replaced use of util with component everywhere
Fixed styling of date time input
Refactored disabling set status button in custom status modal
